python/lsst/meas/astrom/matchPessimisticB.py

```diff
@@ -469,26 +475,102 @@ def _doMatch(self, refCat, sourceCat, wcs, refFluxField, numUsableSources,
                 # in the struct and exit.
                 match_tolerance.maxShift = afwgeom.Angle(matcher_struct.shift,
                                                          afwgeom.arcseconds)
-                match_tolerance.lastMatchedPattern = matcher_struct.pattern_idx
+                match_tolerance.lastMatchedPattern = \
+                    matcher_struct.pattern_idx
                 break
 
+        # The matcher returns all the nearest neighbors that agree between
+        # the reference and source catalog. For the currrent astrometry solver
+        # we need to remove as many false positives as possible before sending
+        # the matches off to the astronomery solver.
+        distances_arcsec = np.rad2deg(matcher_struct.distances_rad) * 3600
+        # Rather than have a hard cut on the max distance we allow for
+        # objects we preform a iterative sigma clipping. We input a
+        # starting point for the clipping based on the largest of the
+        # distances we have. The hope is this distance will not be so
+        # large as to have the sigma_clip return a large unphysical value
+        # and not so small that we don't allow for some distortion in the
+        # WCS between source and reference.
+        clipped_dist_arcsec = self._sigma_clip(
+            values=distances_arcsec, n_sigma=2,
+            input_cut=np.max((10 * wcs.pixelScale().asArcseconds(),
+                              maxMatchDistArcSec,
+                              maxShiftArcseconds)))
+        # We pick the largest of the sigma clipping or 2 pixels. This pevents
+        # the AstrometryTask._matchAndFitWCS from overfitting to a small
+        # number of objects and also allows the WCS fitter to use more
+        # matches as the WCS fit improves.
+        dist_cut_arcsec = np.max(
+            (2 * wcs.pixelScale().asArcseconds(),
+             clipped_dist_arcsec))
+
         # A match has been found, return our list of matches and
         # return.
         matches = []
-        for match_ids, dist in zip(matcher_struct.matches,
-                                   matcher_struct.distances):
-            match = afwTable.ReferenceMatch()
-            match.first = refCat[match_ids[1]]
-            match.second = sourceCat[match_ids[0]]
-            match.distance = match.first.getCoord().angularSeparation(
-                match.second.getCoord())
-            matches.append(match)
+        for match_id_pair, dist_arcsec in zip(matcher_struct.match_ids,
+                                              distances_arcsec):
+            if dist_arcsec < dist_cut_arcsec:
+                match = afwTable.ReferenceMatch()
+                match.first = refCat[match_id_pair[1]]
+                match.second = sourceCat[match_id_pair[0]]
+                # We compute the true distance along and sphere instead
+                # and store it as an afw.geom.Angle object. The previous
+                # distances we used were aproximate.
```

Comment:
Unfortunately you can't store the distance as an `Angle` because `ReferenceMath.distance` is a C++ double. The angle is almost certainly being silently cast to radians; please make the cast explicit (and update the code comment), e.g.:
```
match.distance = match.first.getCoord().angularSeparation(match.second.getCoord()).asRadians()
```